smart-contract/program/src/state.rs

```diff
@@ -250,36 +243,74 @@ impl StakePoolHeader {
         Ok(Self {
             tag: Tag::InactiveStakePool as u8,
             total_staked: 0,
+            total_staked_delta: 0,
             current_day_idx: 0,
             _padding: [0; 4],
-            last_crank_time: Clock::get()?.unix_timestamp,
-            last_claimed_time: Clock::get()?.unix_timestamp,
+            last_claimed_offset: 0,
             owner: owner.to_bytes(),
             nonce,
             vault: vault.to_bytes(),
             minimum_stake_amount,
             stakers_part: STAKER_MULTIPLIER,
-            unstake_period: UNSTAKE_PERIOD,
+            last_delta_update_offset: 0,
         })
     }
 
     pub fn close(&mut self) {
         self.tag = Tag::Deleted as u8
     }
 
-    pub fn deposit(&mut self, amount: u64) -> ProgramResult {
+    pub fn deposit(
+        &mut self,
+        amount: u64,
+        system_snapshot_offset: i64,
+        contract_create_time: i64,
+    ) -> ProgramResult {
         self.total_staked = self
             .total_staked
             .checked_add(amount)
             .ok_or(AccessError::Overflow)?;
-        Ok(())
+        self.update_delta(amount as i64, system_snapshot_offset, contract_create_time)
     }
 
-    pub fn withdraw(&mut self, amount: u64) -> ProgramResult {
+    pub fn withdraw(
+        &mut self,
+        amount: u64,
+        system_snapshot_offset: i64,
+        contract_create_time: i64,
+    ) -> ProgramResult {
         self.total_staked = self
             .total_staked
             .checked_sub(amount)
             .ok_or(AccessError::Overflow)?;
+        self.update_delta(
+            -(amount as i64),
+            system_snapshot_offset,
+            contract_create_time,
+        )
+    }
+
+    // private
+    fn update_delta(
```

Comment:
I think `total_staked_delta` and `last_delta_update_offset` should be reset in the `crank` instruction. And `stake` and `unstake` should error if the pool and the central state are out of sync.

And then you move this directly into `deposit` and `withdraw`
```rust
        self.total_staked_delta = self
            .total_staked_delta
            .checked_add/sub(amount)
            .ok_or(AccessError::Overflow)?;
        Ok(())
```

Reply:
I disagree with this. With the current implementation it is possible to stake/unstake even when the pool has not been cranked on this day which is a lot less restrictive on the users